### ISSUE TITLE
Access to @RequestHeader

### ISSUE THREAD

**Issue by mehdijouan** [phoenixnap/springmvc-raml-plugin#64]
Hi,
For a given case I need to get headers information from a request. Spring has defined a parameter annotation named **@RequestHeader** that can be used as following:

```java
public ResponseEntity<com.test.samples.model.SampleUnit> getSampleUnitById(
        @PathVariable String id, @RequestHeader HttpHeaders headers) {
```
Is there a way to have such parameter, maybe a rule or a maven property to activate ?

**Comment by kurtpa** [comment]
Yes, we should support this.  In RAML it is possible to define headers,  so it should be trivial to extract them

**Comment by aweisser** [comment]
We currently have a similar issue in our project. 
We'd like to add an Accept-Language Header to the controllers' parameter list like this:
`@RequestHeader("Accept-Language") String acceptLanguage`
This would lead to a more specific and explicit method signatur than the generic `@RequestHeader HttpHeaders headers`.

Another approach could be to just provide a configuration flag that adds the generic `@HttpHeader HttpHeaders headers` to every controller method.

A third approach could be to create the generic `@HttpHeader HttpHeaders headers` whenever a `headers` field is specified in .raml.

What do you think?

**Comment by aweisser** [comment]
My thoughts:

The .raml spec describes the `headers` field as a way to specify **non-standard HTTP Headers**.
(See [RAML 0.8 SPEC](https://github.com/raml-org/raml-spec/blob/master/versions/raml-08/raml-08.md#headers)).

So imho the explicit apporach (like `@RequestHeader("Accept-Language") String acceptLanguage`) should be used for this because this way the controller method signature reflects the .raml in a more accurate way.

The generic approach (like `@HttpHeader HttpHeaders headers`) should be used if the use wants to access the standard HTTP Headers. This one should be configurable via a switch a it should be mandatory for ALL controller methods, if turned on.

**Comment by kurtpa** [comment]
I started writing this before your second response :D

I agree entirely - 
I think that the first approach you mention, ie having these as explicit variables in the signature would be most useful for headers which mean something to the organisation using them. Sometimes REST being what it is, some people might want to use standard headers for something or other and this approach would allow for that too. ie if they specify it in the raml, it will show up in the controller/delegate method parameters.

However i like also giving the option of enabling generic headers in the controller generation - and i agree, this should be an always on/off thing. The only issue i have with this is that the interface will also be polluted with the HttpHeaders. An alternate approach could be to offer a generic switch for the HttpServletRequest which could then be used for headers as well as any other corner case which would crop up.

Another option for this would be to create a filter which wraps around the controllers, which stores the HttpServletRequest in a wrapped ThreadLocal and then clears it on exit. This is a tiny bit more memory intensive, less clean but would result in cleaner interfaces.

Either way i think we should tackle these 2 as seperate features

Thoughts?

**Comment by mehdijouan** [comment]
I also agree with @aweisser. 
We are defining our own headers in the raml. So as a first step, I'm more in favor to have each header defined as a parameter of the method.

Having the full HttpHeaders object could be interesting and I prefer to have it in the method stub as it is the way spring defined it. 
When activating it, do you intend to replace all custom headers with one default HttpHeaders object ?

**Comment by kurtpa** [comment]
For the full object,  either HttpHeaders or HttpServletRequest i would think it should show up together with the individual headers as parameters

**Comment by kurtpa** [comment]
@mehdijouan, there is a PR for this functionality from @tofcoder which i think will satisfy the requirements

**Comment by kurtpa** [comment]
PR Merged! Thank you everyone

**Comment by mehdijouan** [comment]
Hi, do you plan a new release soon ? :smile:

**Comment by kurtpa** [comment]
doing it right now :)

**Comment by mehdijouan** [comment]
Perfect!

**Comment by kurtpa** [comment]
Released 0.7.1 to central, should sync within the next 2 hrs
